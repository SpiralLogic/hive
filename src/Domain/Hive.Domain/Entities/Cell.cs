using System;
using System.Collections.Generic;
using System.Linq;

namespace Hive.Domain.Entities
{
    public class Cell : IEquatable<Cell>
    {
        public Coords Coords { get; init; }
        public Stack<Tile> Tiles { get; init; } = new Stack<Tile>();
     
        public Cell(Coords coords)
        {
            Coords = coords;
        }

        public Cell AddTile(Tile tile)
        {
            Tiles.Push(tile);
            return this;
        }

        public bool IsEmpty() => !Tiles.Any();

        public Tile TopTile() => Tiles.Peek();

        public Tile RemoveTopTile() => Tiles.Pop();

        public bool Equals(Cell? other)
        {
            if (other is null) return false;
            if (ReferenceEquals(this, other)) return true;
            return Coords.Equals(other.Coords);
        }

        public override bool Equals(object? obj)
        {
            if (obj is null) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != GetType()) return false;
            return Equals((Cell)obj);
        }

        public override int GetHashCode()
        {
            return Coords.GetHashCode();
        }
    }
}