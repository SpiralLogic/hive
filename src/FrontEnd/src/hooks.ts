import { HiveEvent, HiveEventEmitter, HiveEventListener } from './hive-event-emitter';
import { Inputs, useEffect, useReducer } from 'preact/hooks';

const hiveEventEmitter = new HiveEventEmitter();

export const useHiveEventEmitter = (
  hiveEventHandler?: HiveEventListener<HiveEvent>,
  inputs?: Inputs
): HiveEventEmitter => {
  if (hiveEventHandler)
    useEffect(() => {
      hiveEventEmitter.add(hiveEventHandler);
      return () => {
        hiveEventEmitter.remove(hiveEventHandler);
      };
    }, inputs);

  return hiveEventEmitter;
};

const classReducer = (
  initialClasses: Array<string>,
  action: { type: 'add' | 'remove'; classes: string[] }
): string[] => {
  const { type, classes } = action;
  switch (type) {
    case 'add':
      return [...initialClasses, ...classes.filter((c) => !initialClasses.includes(c))];
    case 'remove':
      return initialClasses.filter((c) => !classes.includes(c));
  }
};

export const useClassReducer = (
  initialClasses: string[]
): [Array<string>, (action: { type: 'add' | 'remove'; classes: string[] }) => void] =>
  useReducer(classReducer, initialClasses);
