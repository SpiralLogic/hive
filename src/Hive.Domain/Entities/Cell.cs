using System.Collections.Generic;
using System.Linq;

namespace Hive.Domain.Entities;

public sealed record Cell(Coords Coords)
{
    public Stack<Tile> Tiles { get; init; } = new();
    private readonly int _hashCode = Coords.GetHashCode();

    public bool Equals(Cell? other)
    {
        if (ReferenceEquals(other, null)) return false;
        return ReferenceEquals(this, other) || Coords.Equals(other.Coords);
    }

    public Cell AddTile(Tile tile)
    {
        Tiles.Push(tile);
        return this;
    }

    public bool IsEmpty()
    {
        return !Tiles.Any();
    }

    public Tile TopTile()
    {
        return Tiles.Peek();
    }

    public Tile RemoveTopTile()
    {
        return Tiles.Pop();
    }

    public override int GetHashCode() =>
        _hashCode;
}
