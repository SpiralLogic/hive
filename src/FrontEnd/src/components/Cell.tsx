import {Cell, HexCoordinates, Tile as TileType} from '../domain';
import {FunctionComponent, h} from 'preact';
import {TileDragEvent, useCellDropEmitter, useTileDragEmitter,} from '../emitters';
import {deepEqual} from 'fast-equals';
import {handleDragOver} from '../handlers';
import {memo} from 'preact/compat';
import {useEffect, useState} from 'preact/hooks';
import Tile from './Tile';

type Props = Cell;

const CellFC: FunctionComponent<Props> = (props: Props) => {
    const {tiles, coords} = props;
    const [tileDragEmitter, cellDropEmitter] = [
        useTileDragEmitter(),
        useCellDropEmitter(),
    ];
    const isValidMove = (validMoves: HexCoordinates[]) =>
        validMoves.some((dest) => deepEqual(coords, dest));
    const [classes, setClasses] = useState(['hex', 'cell']);
    const [currentTile, setCurrentTile] = useState<TileType | null>(null);

    function handleDragLeave(ev: { stopPropagation: () => void }) {
        ev.stopPropagation();
        if (currentTile)    setClasses(classes.filter(e => e !== 'active'))
    }

    function handleDragEnter(ev: { stopPropagation: () => void }) {
        ev.stopPropagation();
        if (currentTile) setClasses([...classes, 'active']);
    }

    function handleClickEvent(ev: { stopPropagation: () => void }) {
        if (currentTile) {
            setClasses([...classes, 'active']);
            tileDragEmitter.emit({type: 'end', tile: currentTile})
        }
    }

    const handleTileEvent = (e: TileDragEvent) => {
        const valid = isValidMove(e.tile.moves);
        if (e.type === 'start' && valid) {
            setClasses([...classes, 'can-drop']);
            setCurrentTile(e.tile)
        } else if (e.type === "start") {
            setCurrentTile(e.tile)
        } else if (e.type === 'end') {
            if (valid && currentTile && classes.includes('active'))
                cellDropEmitter.emit({
                    type: 'drop',
                    move: {coords, tileId: currentTile.id},
                });
            setCurrentTile(null);
            setClasses(['hex', 'cell']);
        }
    };

    useEffect(() => {
        tileDragEmitter.add(handleTileEvent);
        return () => tileDragEmitter.remove(handleTileEvent);
    });

    const attributes = {
        className: classes.join(' '),
        ondragover: handleDragOver,
        ondragleave: handleDragLeave,
        ondragenter: handleDragEnter,
        onmouseenter: handleDragEnter,
        onmouseleave: handleDragLeave,
        onclick: handleClickEvent,
    };

    return (
        <div {...attributes}>
            {tiles.length > 0 && <Tile {...tiles[0]} />}
        </div>
    );
};

CellFC.displayName = 'Cell';
export default memo(
    CellFC,
    (p, n) => deepEqual(p.coords, n.coords) && !p.tiles.length && !n.tiles.length
);
