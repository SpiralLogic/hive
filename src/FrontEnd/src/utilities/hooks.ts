import { HiveDispatcher, HiveEventListener, HiveIntent } from './hive-dispatcher';
import { useEffect, useReducer } from 'preact/hooks';

const hiveDispatcher = new HiveDispatcher();

export const useHiveDispatcher = (): HiveDispatcher => {
  return hiveDispatcher;
};

export const addHiveDispatchListener = <T extends HiveIntent>(
  type: T['type'],
  listener: HiveEventListener<T>
): void => {
  useEffect(() => {
    const emitter = useHiveDispatcher();
    return emitter.add<T>(type, listener);
  });
};

export const dispatchHiveEvent = hiveDispatcher.dispatch;

const classReducer = (
  initialClasses: string,
  action: { type: 'add' | 'remove'; classes: string[] }
): string => {
  const classList = new Set(initialClasses.split(' '));
  if (action.type === `add`) {
    action.classes.forEach((c) => classList.add(c));
  } else if (action.type === 'remove') {
    action.classes.forEach((c) => classList.delete(c));
  }
  classList.delete('');
  return Array.from(classList).join(' ');
};

export const useClassReducer = (
  initialClasses: string
): [string, (action: { type: 'add' | 'remove'; classes: string[] }) => void] =>
  useReducer(classReducer, initialClasses);
