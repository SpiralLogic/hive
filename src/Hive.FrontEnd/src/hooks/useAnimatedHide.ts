import {useEffect} from "preact/hooks";
import {useClassSignal} from "./useClassSignal";

export const useAnimatedHide = (initialClasses:string, shouldHide: boolean) => {
    const [classes, action] = useClassSignal(initialClasses);
    useEffect(() => {
        if (shouldHide) {
            setTimeout(() => action.add('hide'), 100);
        }
    }, [shouldHide, action]);

    const ontransitionend = () => {
        if (classes.peek().includes('hide')) action.add('hidden');
    };

    return {class:classes, onTransitionEnd: ontransitionend, onAnimationEnd: ontransitionend} as const;
}
export const useHide = (shouldHide: boolean, [, action]: ReturnType<typeof useClassSignal>) => {
    useEffect(() => (shouldHide ? action.add('hide') : action.remove('hide')), [shouldHide, action]);

}