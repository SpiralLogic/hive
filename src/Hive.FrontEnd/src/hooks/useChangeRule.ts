import {useCallback, useMemo, useReducer} from "preact/hooks";
import {Rule} from "../components/rules/rule";

export const useChangeRule = (ruleList: Rule[]) => {
    const changeRule = useCallback((currentRuleIndex: number, {type}: { type: 'next' | 'prev' }): number => {
        if (type === 'next') return (currentRuleIndex + 1) % ruleList.length;
        if (currentRuleIndex > 0) return (currentRuleIndex - 1) % ruleList.length;
        return ruleList.length - 1;
    }, [ruleList]);

    const [currentRuleIndex, changeRuleDispatcher] = useReducer(changeRule, 0);
    const handlers = useMemo(() => ({
        previous: () => changeRuleDispatcher({type: 'prev'}),
        next: () => changeRuleDispatcher({type: 'next'})
    }) as const, []);

    return [ruleList[currentRuleIndex], handlers] as const;
}