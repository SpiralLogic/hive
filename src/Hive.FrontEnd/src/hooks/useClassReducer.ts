import {  useMemo } from 'preact/hooks';
import { useSignal } from '@preact/signals';

const getClassString = (classList: Set<string>) => Array.from(classList).join(' ');
export const useClassSignal = (...initialClasses: string[]) => {
  const classList = new Set<string>(initialClasses.map((c) => c.trim()));
  const classSignal = useSignal(getClassString(classList));
  const classActions = useMemo(
    () => ({
      add: (...classes: string[]) => {
        classes.forEach((c) => classList.add(c));
        classSignal.value = Array.from(classList).join(' ');
      },
      remove: (...classes: string[]) => {
        classes.forEach((c) => classList.delete(c));
        classSignal.value = getClassString(classList);
      },
    }),
    []
  );
  return [classSignal, classActions] as const;
};
