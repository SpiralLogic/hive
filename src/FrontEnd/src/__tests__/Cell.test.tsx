import { act, fireEvent, render } from '@testing-library/preact';
import { deepEqual } from 'fast-equals';
import { h } from 'preact';
import { renderElement, simulateEvent } from './helpers';
import { useHiveEventEmitter } from '../hooks';
import Cell from '../components/Cell';

jest.mock('fast-equals', () => ({ deepEqual: jest.fn(() => true) }));
describe('Cell Tests', () => {
  const moveTileSpy = jest.fn();

  const movingTile = {
    id: 2,
    moves: [
      { q: 0, r: 0 },
      { q: 2, r: 2 },
    ],
    creature: '',
    playerId: 1,
  };

  const createCellWithNoTile = () => {
    const cell = { coords: { q: 0, r: 0 }, tiles: [] };
    return renderElement(<Cell {...cell} />);
  };

  const createCellWithTile = () => {
    const tile = { id: 2, playerId: 1, creature: 'fly', moves: [] };
    const cell = { coords: { q: 1, r: 1 }, tiles: [tile] };

    return renderElement(<Cell {...cell} />);
  };

  const createCellWithTileAndDrop = () => {
    const tile = { id: 2, playerId: 1, creature: 'ant', moves: [{ r: 0, q: 0 }] };
    const cell = { coords: { q: 2, r: 2 }, tiles: [tile] };

    return renderElement(<Cell {...cell} />);
  };

  const createCellNoDrop = () => {
    const cell = { coords: { q: 6, r: 6 }, tiles: [] };
    return renderElement(<Cell {...cell} />);
  };

  const createCellCanDrop = createCellWithNoTile;
  const createCellWithTileNoDrop = createCellWithTile;

  describe('Cell render', () => {
    test('has default classes', () => {
      expect(createCellWithTile()).toHaveClass('hex cell');
      expect(createCellWithNoTile()).toHaveClass('hex cell');
    });

    test(`top tile is rendered when tiles isn't empty`, () => {
      const tiles = createCellWithTile().getElementsByClassName('tile');
      expect(tiles).toHaveLength(1);
    });

    test('component is memorized with deep equal', () => {
      const props = { coords: { q: 0, r: 0 }, tiles: [] };
      const cell = <Cell {...props} />;
      render(cell).rerender(cell);
      expect(deepEqual).toHaveBeenCalledTimes(1);
    });
  });

  describe('drag and drop', () => {
    const emitter = useHiveEventEmitter();

    function emitHiveEvent(type: 'tileSelected' | 'tileDropped') {
      act(() =>
        emitter.emit({
          type,
          tile: movingTile,
        })
      );
    }

    test('dragover allows drop', () => {
      const preventDefault = simulateEvent(createCellWithTile(), 'dragover');
      expect(preventDefault).toHaveBeenCalled();
    });

    test('cell is available on drag start', () => {
      const cellWithTile = createCellWithTileAndDrop();
      const emptyCell = createCellCanDrop();
      emitHiveEvent('tileSelected');

      expect(cellWithTile).toHaveClass('can-drop');
      expect(emptyCell).toHaveClass('can-drop');
    });

    test('available cell is active on tile drag enter', () => {
      const cellWithTile = createCellWithTileAndDrop();
      const emptyCell = createCellCanDrop();
      emitHiveEvent('tileSelected');
      fireEvent.dragEnter(cellWithTile);
      fireEvent.dragEnter(emptyCell);

      expect(cellWithTile).toHaveClass('active');
      expect(emptyCell).toHaveClass('active');
    });

    test('move calls moves tile when cell is valid and active', () => {
      jest.spyOn(useHiveEventEmitter(), 'emit');
      const cellWithTile = createCellWithTileAndDrop();
      const emptyCell = createCellCanDrop();
      emitHiveEvent('tileSelected');
      fireEvent.dragEnter(cellWithTile);
      fireEvent.dragEnter(emptyCell);
      emitHiveEvent('tileDropped');

      expect(useHiveEventEmitter().emit).toHaveBeenCalledWith({
        type: 'move',
        move: { tileId: 2, coords: { q: 0, r: 0 } },
      });

      expect(useHiveEventEmitter().emit).toHaveBeenCalledWith({
        type: 'move',
        move: { tileId: 2, coords: { q: 2, r: 2 } },
      });
    });

    test(`drop doesn't call move tile when cell doesn't allow drop`, () => {
      jest.spyOn(useHiveEventEmitter(), 'emit');
      createCellWithTile();
      createCellNoDrop();
      emitHiveEvent('tileSelected');
      emitHiveEvent('tileDropped');

      expect(useHiveEventEmitter().emit).not.toHaveBeenCalledWith({ type: 'move' });
    });

    test(`invalid cells don't call move tile on drop`, () => {
      createCellWithTileNoDrop();
      createCellNoDrop();
      document.querySelectorAll('.cell').forEach((c) => fireEvent.dragEnter(c));

      expect(moveTileSpy).not.toHaveBeenCalled();
    });

    test('active classes are removed on drag leave', () => {
      createCellWithTile();
      createCellWithNoTile();
      createCellWithTileNoDrop();
      createCellNoDrop();
      emitHiveEvent('tileSelected');
      document.querySelectorAll('.cell').forEach((c) => fireEvent.dragEnter(c));
      document.querySelectorAll('.cell').forEach((c) => fireEvent.dragLeave(c));

      expect(document.getElementsByClassName('active')).toHaveLength(0);
    });

    test('active and can-drop classes are removed on drop', () => {
      createCellWithTile();
      createCellWithNoTile();
      createCellWithTileNoDrop();
      createCellNoDrop();
      emitHiveEvent('tileSelected');
      document.querySelectorAll('.cell').forEach((c) => fireEvent.dragEnter(c));
      emitHiveEvent('tileDropped');

      expect(document.getElementsByClassName('active')).toHaveLength(0);
      expect(document.getElementsByClassName('can-drop')).toHaveLength(0);
    });

    test(`occupied cell with no active tile doesn't stop event propagation'`, () => {
      jest.spyOn(useHiveEventEmitter(), 'emit');
      const cellWithTile = createCellWithTileAndDrop();
      fireEvent.click(cellWithTile);

      expect(useHiveEventEmitter().emit).not.toBeCalled();
    });

    test(`cell click with active tile makes a move`, () => {
      jest.spyOn(useHiveEventEmitter(), 'emit');
      const emptyCell = createCellCanDrop();
      emitHiveEvent('tileSelected');
      fireEvent.click(emptyCell);

      expect(useHiveEventEmitter().emit).toHaveBeenCalledWith({
        type: 'move',
        move: { tileId: 2, coords: { q: 0, r: 0 } },
      });
      expect(useHiveEventEmitter().emit).toBeCalled();
    });

    test(`cell click with no active tile shouldn't move`, () => {
      jest.spyOn(useHiveEventEmitter(), 'emit');
      const emptyCell = createCellNoDrop();
      fireEvent.click(emptyCell);

      expect(useHiveEventEmitter().emit).not.toHaveBeenCalledWith();
    });

    test(`cell click with invalid tile shouldn't move`, () => {
      jest.spyOn(useHiveEventEmitter(), 'emit');
      const emptyCell = createCellNoDrop();
      emitHiveEvent('tileSelected');
      fireEvent.click(emptyCell);

      expect(useHiveEventEmitter().emit).not.toHaveBeenCalledWith();
    });

    test(`enter fires emit event on keydown enter`, () => {
      jest.spyOn(useHiveEventEmitter(), 'emit');
      const emptyCell = createCellCanDrop();
      emitHiveEvent('tileSelected');
      fireEvent.keyDown(emptyCell, { key: 'Enter' });

      expect(useHiveEventEmitter().emit).toHaveBeenCalledWith(expect.objectContaining({ type: 'move' }));
    });

    test(`space fires emit event on keydown enter`, () => {
      jest.spyOn(useHiveEventEmitter(), 'emit');
      const emptyCell = createCellCanDrop();
      emitHiveEvent('tileSelected');
      fireEvent.keyDown(emptyCell, { key: ' ' });

      expect(useHiveEventEmitter().emit).toHaveBeenCalledWith(expect.objectContaining({ type: 'move' }));
    });

    test(`other keys dont emits tile start event`, () => {
      jest.spyOn(useHiveEventEmitter(), 'emit');
      fireEvent.keyDown(createCellCanDrop(), { key: 'Tab' });

      expect(useHiveEventEmitter().emit).not.toHaveBeenCalled();
    });
  });

  describe('Cell Snapshot', () => {
    test('cell with tile matches current snapshot', () => {
      expect(createCellWithTile()).toMatchSnapshot();
    });

    test('cell with no tile matches current snapshot', () => {
      expect(createCellWithNoTile()).toMatchSnapshot();
    });
  });
});
