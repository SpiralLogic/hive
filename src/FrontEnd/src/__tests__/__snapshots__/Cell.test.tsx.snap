// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell Tests Cell Snapshot cell with no tile matches current snapshot 1`] = `
<div
  class="hex cell"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
  </svg>
</div>
`;

exports[`Cell Tests Cell Snapshot cell with tile matches current snapshot 1`] = `
<div
  class="hex cell"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
  </svg>
  <div
    class="player1 hex tile"
    draggable="false"
    title="fly"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#hex"
      />
      <use
        class="creature"
        href="#fly"
      />
    </svg>
  </div>
</div>
`;
